File: clients/go/memory/mcp_handler.go
package memory

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
)

// MCPHandler returns an http.Handler that exposes the 12 standard MCP tools
// matching the hosted server at https://memory.neo4jlabs.com/mcp.
//
//	client, _ := memory.New(
//	    memory.WithEndpoint("https://memory.neo4jlabs.com/v1"),
//	    memory.WithAPIKey(os.Getenv("MEMORY_API_KEY")),
//	)
//	http.Handle("/mcp/", client.MCPHandler())
func (c *Client) MCPHandler() http.Handler {
	mux := http.NewServeMux()

	mux.HandleFunc("POST /tools/list", func(w http.ResponseWriter, _ *http.Request) {
		writeJSON(w, map[string]interface{}{"tools": MCPTools()})
	})

	mux.HandleFunc("POST /tools/call", func(w http.ResponseWriter, r *http.Request) {
		var req struct {
			Name      string                 `json:"name"`
			Arguments map[string]interface{} `json:"arguments"`
		}
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			writeError(w, http.StatusBadRequest, err.Error())
			return
		}
		result, err := MCPDispatch(r.Context(), c, req.Name, req.Arguments)
		if err != nil {
			writeError(w, http.StatusInternalServerError, err.Error())
			return
		}
		writeJSON(w, result)
	})

	return mux
}

// MCPToolDefinition is a JSON Schema-compatible MCP tool descriptor.
type MCPToolDefinition struct {
	Name        string                 `json:"name"`
	Description string                 `json:"description"`
	InputSchema map[string]interface{} `json:"inputSchema"`
}

// MCPTools returns the 12 standard memory tools matching memory.neo4jlabs.com/mcp.
func MCPTools() []MCPToolDefinition {
	prop := func(t string, desc string) map[string]interface{} {
		return map[string]interface{}{"type": t, "description": desc}
	}
	obj := func(props map[string]interface{}, required ...string) map[string]interface{} {
		s := map[string]interface{}{"type": "object", "properties": props}
		if len(required) > 0 {
			s["required"] = required
		}
		return s
	}

	return []MCPToolDefinition{
		{Name: "memory_create_conversation", Description: "Create a new conversation session for a user.",
			InputSchema: obj(map[string]interface{}{
				"user_id":  prop("string", "User identifier"),
				"metadata": map[string]interface{}{"type": "object"},
			}, "user_id")},
		{Name: "memory_add_messages", Description: "Append one or more messages to a conversation.",
			InputSchema: obj(map[string]interface{}{
				"conversation_id": prop("string", "Conversation id"),
				"messages":        map[string]interface{}{"type": "array"},
			}, "conversation_id", "messages")},
		{Name: "memory_get_context", Description: "Three-tier context (reflections + observations + recent messages).",
			InputSchema: obj(map[string]interface{}{
				"conversation_id": prop("string", "Conversation id"),
			}, "conversation_id")},
		{Name: "memory_search_messages", Description: "Search messages within a conversation.",
			InputSchema: obj(map[string]interface{}{
				"conversation_id": prop("string", "Conversation id"),
				"query":           prop("string", "Search query"),
				"limit":           prop("number", "Max results"),
			}, "conversation_id", "query")},
		{Name: "memory_search_entities", Description: "Search the knowledge graph for entities.",
			InputSchema: obj(map[string]interface{}{
				"query": prop("string", "Search query"),
				"type":  prop("string", "Entity type filter"),
				"limit": prop("number", "Max results"),
			}, "query")},
		{Name: "memory_get_entity", Description: "Fetch one entity (with relationships) by id.",
			InputSchema: obj(map[string]interface{}{
				"entity_id": prop("string", "Entity id"),
			}, "entity_id")},
		{Name: "memory_add_entity", Description: "Manually create an entity.",
			InputSchema: obj(map[string]interface{}{
				"name":        prop("string", "Entity name"),
				"type":        prop("string", "Entity type"),
				"description": prop("string", "Description"),
			}, "name", "type")},
		{Name: "memory_get_entity_history", Description: "All conversations that mentioned this entity.",
			InputSchema: obj(map[string]interface{}{
				"entity_id": prop("string", "Entity id"),
			}, "entity_id")},
		{Name: "memory_record_step", Description: "Log a reasoning step under a conversation.",
			InputSchema: obj(map[string]interface{}{
				"conversation_id": prop("string", "Conversation id"),
				"reasoning":       prop("string", "Reasoning text"),
				"action_taken":    prop("string", "Action description"),
				"result":          prop("string", "Optional result"),
			}, "conversation_id", "reasoning", "action_taken")},
		{Name: "memory_record_tool_call", Description: "Log a tool invocation tied to a reasoning step.",
			InputSchema: obj(map[string]interface{}{
				"step_id":     prop("string", "Step id"),
				"tool_name":   prop("string", "Tool name"),
				"input":       prop("string", "Input as JSON string"),
				"output":      prop("string", "Output as JSON string"),
				"status":      prop("string", "success | error | timeout"),
				"duration_ms": prop("number", "Duration ms"),
			}, "tool_name", "status")},
		{Name: "memory_get_trace", Description: "Full reasoning trace for a conversation.",
			InputSchema: obj(map[string]interface{}{
				"conversation_id": prop("string", "Conversation id"),
			}, "conversation_id")},
		{Name: "memory_explain_decision", Description: "Detailed explanation of one reasoning step.",
			InputSchema: obj(map[string]interface{}{
				"step_id": prop("string", "Step id"),
			}, "step_id")},
	}
}

// MCPDispatch executes one of the 12 standard MCP tools against a Client.
func MCPDispatch(ctx context.Context, c *Client, name string, args map[string]interface{}) (interface{}, error) {
	switch name {
	case "memory_create_conversation":
		return c.ShortTerm.CreateConversation(ctx, CreateConversationParams{
			UserID:   getStr(args, "user_id"),
			Metadata: getMap(args, "metadata"),
		})
	case "memory_add_messages":
		conversationID := getStr(args, "conversation_id")
		raw, _ := args["messages"].([]interface{})
		msgs := make([]BulkMessageInput, 0, len(raw))
		for _, item := range raw {
			m, _ := item.(map[string]interface{})
			if m == nil {
				continue
			}
			msgs = append(msgs, BulkMessageInput{
				Role:     MessageRole(getStr(m, "role")),
				Content:  getStr(m, "content"),
				Metadata: getMap(m, "metadata"),
			})
		}
		if len(msgs) == 1 {
			opts := []func(*AddMessageParams){}
			if msgs[0].Metadata != nil {
				opts = append(opts, WithMetadata(msgs[0].Metadata))
			}
			msg, err := c.ShortTerm.AddMessage(ctx, conversationID, msgs[0].Role, msgs[0].Content, opts...)
			if err != nil {
				return nil, err
			}
			return []*Message{msg}, nil
		}
		return c.ShortTerm.BulkAddMessages(ctx, conversationID, msgs)
	case "memory_get_context":
		return c.ShortTerm.GetContext(ctx, getStr(args, "conversation_id"))
	case "memory_search_messages":
		return c.ShortTerm.SearchMessages(ctx, getStr(args, "query"),
			WithSessionID(getStr(args, "conversation_id")),
			WithSearchLimit(getInt(args, "limit", 10)),
		)
	case "memory_search_entities":
		return c.LongTerm.SearchEntities(ctx, getStr(args, "query"), getInt(args, "limit", 10))
	case "memory_get_entity":
		return c.LongTerm.GetEntity(ctx, getStr(args, "entity_id"))
	case "memory_add_entity":
		opts := []func(*AddEntityParams){}
		if d := getStr(args, "description"); d != "" {
			opts = append(opts, WithDescription(d))
		}
		return c.LongTerm.AddEntity(ctx, getStr(args, "name"), getStr(args, "type"), opts...)
	case "memory_get_entity_history":
		return c.LongTerm.GetEntityHistory(ctx, getStr(args, "entity_id"))
	case "memory_record_step":
		return c.Reasoning.RecordStep(ctx, RecordStepInput{
			ConversationID: getStr(args, "conversation_id"),
			Reasoning:      getStr(args, "reasoning"),
			ActionTaken:    getStr(args, "action_taken"),
			Result:         getStr(args, "result"),
		})
	case "memory_record_tool_call":
		stepID := getStr(args, "step_id")
		toolName := getStr(args, "tool_name")
		argsMap := map[string]interface{}{"input": getStr(args, "input")}
		opts := []func(*RecordToolCallParams){
			WithStatus(ToolCallStatus(getStr(args, "status"))),
		}
		if out := getStr(args, "output"); out != "" {
			opts = append(opts, WithResult(out))
		}
		if d := getInt(args, "duration_ms", 0); d > 0 {
			opts = append(opts, WithDurationMs(d))
		}
		return c.Reasoning.RecordToolCall(ctx, stepID, toolName, argsMap, opts...)
	case "memory_get_trace":
		return c.Reasoning.GetTraceByConversation(ctx, getStr(args, "conversation_id"))
	case "memory_explain_decision":
		return c.Reasoning.ExplainStep(ctx, getStr(args, "step_id"))
	default:
		return nil, fmt.Errorf("unknown memory tool: %s", name)
	}
}

func getStr(m map[string]interface{}, key string) string {
	v, ok := m[key]
	if !ok {
		return ""
	}
	s, ok := v.(string)
	if !ok {
		return ""
	}
	return s
}

func getMap(m map[string]interface{}, key string) map[string]interface{} {
	v, ok := m[key].(map[string]interface{})
	if !ok {
		return nil
	}
	return v
}

func getInt(m map[string]interface{}, key string, def int) int {
	v, ok := m[key].(float64)
	if !ok {
		return def
	}
	return int(v)
}

func writeJSON(w http.ResponseWriter, data interface{}) {
	w.Header().Set("Content-Type", "application/json")
	_ = json.NewEncoder(w).Encode(data)
}

func writeError(w http.ResponseWriter, code int, msg string) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(code)
	_ = json.NewEncoder(w).Encode(map[string]string{"error": msg})
}

File: clients/csharp/src/Neo4j.AgentMemory/Mcp/McpHandler.cs
using System.Text.Json;
using Neo4j.AgentMemory.Models;

namespace Neo4j.AgentMemory.Mcp;

/// <summary>JSON Schema-compatible MCP tool descriptor.</summary>
public class McpToolDefinition
{
    public string Name { get; set; } = "";
    public string Description { get; set; } = "";
    public Dictionary<string, object?> InputSchema { get; set; } = new();
}

/// <summary>
/// 12-tool MCP surface matching the hosted server at memory.neo4jlabs.com/mcp.
/// </summary>
public static class McpHandler
{
    public static List<McpToolDefinition> Tools()
    {
        Dictionary<string, object?> Prop(string type, string desc) =>
            new() { ["type"] = type, ["description"] = desc };

        Dictionary<string, object?> ObjSchema(Dictionary<string, object?> props, params string[] required)
        {
            var schema = new Dictionary<string, object?>
            {
                ["type"] = "object",
                ["properties"] = props,
            };
            if (required.Length > 0) schema["required"] = required;
            return schema;
        }

        return new List<McpToolDefinition>
        {
            new() { Name = "memory_create_conversation", Description = "Create a new conversation session for a user.",
                InputSchema = ObjSchema(new() { ["user_id"] = Prop("string", "User identifier"), ["metadata"] = new Dictionary<string, object?> { ["type"] = "object" } }, "user_id") },
            new() { Name = "memory_add_messages", Description = "Append one or more messages to a conversation.",
                InputSchema = ObjSchema(new() { ["conversation_id"] = Prop("string", "Conversation id"), ["messages"] = new Dictionary<string, object?> { ["type"] = "array" } }, "conversation_id", "messages") },
            new() { Name = "memory_get_context", Description = "Three-tier context.",
                InputSchema = ObjSchema(new() { ["conversation_id"] = Prop("string", "Conversation id") }, "conversation_id") },
            new() { Name = "memory_search_messages", Description = "Search messages within a conversation.",
                InputSchema = ObjSchema(new() { ["conversation_id"] = Prop("string", "Conversation id"), ["query"] = Prop("string", "Search query"), ["limit"] = Prop("number", "Max results") }, "conversation_id", "query") },
            new() { Name = "memory_search_entities", Description = "Search the knowledge graph for entities.",
                InputSchema = ObjSchema(new() { ["query"] = Prop("string", "Query"), ["type"] = Prop("string", "Type filter"), ["limit"] = Prop("number", "Max") }, "query") },
            new() { Name = "memory_get_entity", Description = "Fetch one entity (with relationships) by id.",
                InputSchema = ObjSchema(new() { ["entity_id"] = Prop("string", "Entity id") }, "entity_id") },
            new() { Name = "memory_add_entity", Description = "Manually create an entity.",
                InputSchema = ObjSchema(new() { ["name"] = Prop("string", "Name"), ["type"] = Prop("string", "Type"), ["description"] = Prop("string", "Description") }, "name", "type") },
            new() { Name = "memory_get_entity_history", Description = "All conversations that mentioned this entity.",
                InputSchema = ObjSchema(new() { ["entity_id"] = Prop("string", "Entity id") }, "entity_id") },
            new() { Name = "memory_record_step", Description = "Log a reasoning step under a conversation.",
                InputSchema = ObjSchema(new() { ["conversation_id"] = Prop("string", "Conversation id"), ["reasoning"] = Prop("string", "Reasoning"), ["action_taken"] = Prop("string", "Action"), ["result"] = Prop("string", "Result") }, "conversation_id", "reasoning", "action_taken") },
            new() { Name = "memory_record_tool_call", Description = "Log a tool invocation tied to a reasoning step.",
                InputSchema = ObjSchema(new() { ["step_id"] = Prop("string", "Step id"), ["tool_name"] = Prop("string", "Tool name"), ["input"] = Prop("string", "Input"), ["output"] = Prop("string", "Output"), ["status"] = Prop("string", "Status"), ["duration_ms"] = Prop("number", "Duration") }, "tool_name", "status") },
            new() { Name = "memory_get_trace", Description = "Full reasoning trace for a conversation.",
                InputSchema = ObjSchema(new() { ["conversation_id"] = Prop("string", "Conversation id") }, "conversation_id") },
            new() { Name = "memory_explain_decision", Description = "Detailed explanation of one reasoning step.",
                InputSchema = ObjSchema(new() { ["step_id"] = Prop("string", "Step id") }, "step_id") },
        };
    }

    /// <summary>Dispatch one of the 12 standard MCP tool calls to a MemoryClient.</summary>
    public static async Task<object?> DispatchAsync(MemoryClient client, string toolName, Dictionary<string, object?> args, CancellationToken ct = default)
    {
        string GetStr(string key) => args.TryGetValue(key, out var v) && v != null ? v.ToString()! : "";
        int GetInt(string key, int def)
        {
            if (!args.TryGetValue(key, out var v) || v == null) return def;
            if (v is int i) return i;
            if (v is long l) return (int)l;
            if (v is double d) return (int)d;
            if (int.TryParse(v.ToString(), out var parsed)) return parsed;
            return def;
        }
        Dictionary<string, object?>? GetMap(string key)
        {
            if (!args.TryGetValue(key, out var v)) return null;
            return v as Dictionary<string, object?>;
        }

        return toolName switch
        {
            "memory_create_conversation" =>
                await client.ShortTerm.CreateConversationAsync(GetStr("user_id"), GetMap("metadata"), ct),
            "memory_add_messages" => await DispatchAddMessagesAsync(client, GetStr("conversation_id"), args, ct),
            "memory_get_context" =>
                await client.ShortTerm.GetContextAsync(GetStr("conversation_id"), ct),
            "memory_search_messages" =>
                await client.ShortTerm.SearchMessagesAsync(GetStr("query"), GetStr("conversation_id"), GetInt("limit", 10), 0.0, ct),
            "memory_search_entities" =>
                await client.LongTerm.SearchEntitiesAsync(GetStr("query"), GetInt("limit", 10), GetStr("type") is var t && t.Length > 0 ? t : null, ct),
            "memory_get_entity" =>
                await client.LongTerm.GetEntityAsync(GetStr("entity_id"), ct),
            "memory_add_entity" =>
                await client.LongTerm.AddEntityAsync(GetStr("name"), GetStr("type"), GetStr("description") is var d && d.Length > 0 ? d : null, ct),
            "memory_get_entity_history" =>
                await client.LongTerm.GetEntityHistoryAsync(GetStr("entity_id"), ct),
            "memory_record_step" =>
                await client.Reasoning.RecordStepAsync(GetStr("conversation_id"), GetStr("reasoning"), GetStr("action_taken"), GetStr("result") is var r && r.Length > 0 ? r : null, ct),
            "memory_record_tool_call" =>
                await client.Reasoning.RecordToolCallAsync(
                    GetStr("step_id"),
                    GetStr("tool_name"),
                    new Dictionary<string, object?> { ["input"] = GetStr("input") },
                    GetStr("output"),
                    GetStr("status") switch { "error" => ToolCallStatus.Error, "timeout" => ToolCallStatus.Timeout, _ => ToolCallStatus.Success },
                    GetInt("duration_ms", 0) is var ms && ms > 0 ? ms : null,
                    null, ct),
            "memory_get_trace" =>
                await client.Reasoning.GetTraceByConversationAsync(GetStr("conversation_id"), ct),
            "memory_explain_decision" =>
                await client.Reasoning.ExplainStepAsync(GetStr("step_id"), ct),
            _ => throw new ArgumentException($"Unknown memory tool: {toolName}"),
        };
    }

    private static async Task<List<Message>> DispatchAddMessagesAsync(MemoryClient client, string conversationId, Dictionary<string, object?> args, CancellationToken ct)
    {
        if (!args.TryGetValue("messages", out var raw) || raw is not List<object?> rawList || rawList.Count == 0)
            return new List<Message>();
        var messages = new List<BulkMessageInput>();
        foreach (var item in rawList)
        {
            if (item is not Dictionary<string, object?> m) continue;
            messages.Add(new BulkMessageInput
            {
                Role = m.GetValueOrDefault("role")?.ToString() ?? "user",
                Content = m.GetValueOrDefault("content")?.ToString() ?? "",
                Metadata = m.GetValueOrDefault("metadata") as Dictionary<string, object?>,
            });
        }
        if (messages.Count == 1)
        {
            var only = messages[0];
            var role = only.Role switch { "assistant" => MessageRole.Assistant, "system" => MessageRole.System, _ => MessageRole.User };
            return new List<Message> { await client.ShortTerm.AddMessageAsync(conversationId, role, only.Content, only.Metadata, ct) };
        }
        return await client.ShortTerm.BulkAddMessagesAsync(conversationId, messages, ct);
    }

    /// <summary>Convenience — serialize a tool list as MCP-compatible JSON.</summary>
    public static string ToolsJson() => JsonSerializer.Serialize(new { tools = Tools() });
}
